tests/autorestart_test/test_container_autorestart.py

```diff
@@ -0,0 +1,295 @@
+    restarted = wait_until(CONTAINER_RESTART_THRESHOLD_SECS, 
+                  CONTAINER_CHECK_INTERVAL_SECS,
+                  check_container_status, duthost, container_name, False)
+    if not restarted:
+        if is_hiting_start_limit(duthost, container_name):
+            logging.info("{} hits start limit and clear reset-failed...".format(container_name))
+            duthost.shell("sudo systemctl reset-failed {}.service".format(container_name))
+            duthost.shell("sudo systemctl start {}.service".format(container_name))
+            pytest_assert(wait_until(CONTAINER_RESTART_THRESHOLD_SECS, 
+                      CONTAINER_CHECK_INTERVAL_SECS,
+                      check_container_status, duthost, container_name, False), 
+                  "Failed to restart {} after reset-failed is cleared".format(container_name))
+        else:
+            pytest_assert(False, "Failed to restart {}".format(container_name)) 
+    logging.info("{} is restarted".format(container_name))
+
+def verify_no_autorestart_with_non_critical_process(duthost, container_name, program_name,
+                                                    program_status, program_pid):
+    """
+    @summary: Killing a non-critical process in a container to verify whether the container 
+              is still in the running state 
+    """
+    kill_process_by_pid(duthost, container_name, program_name, program_status, program_pid)
+
+    logging.info("Checking whether the {} is still running...".format(container_name))
+    pytest_assert(wait_until(CONTAINER_STOP_THRESHOLD_SECS, 
+                      CONTAINER_CHECK_INTERVAL_SECS,
+                      check_container_status, duthost, container_name, False),
+                  "{} is stopped unexpectedly".format(container_name))
+    logging.info("{} is running".format(container_name))
+
+def test_containers_autorestart(duthost):
+    """
+    @summary: Test the auto-restart feature of each contianer against two scenarios: killing
+              a non-critical process to verity the container is still running; killing each
+              critical process to verify the container will be stopped and restarted
+    """ 
+    container_autorestart_info = get_autorestart_container_and_state(duthost)
+    for container_name in container_autorestart_info:
+        is_running = is_container_running(duthost, container_name)
+        if is_running == "false":
+            pytest_assert(False, "{} is not running".format(container_name))
+
+        logging.info("Start testing the container {}...".format(container_name))
+
+        if container_name == "sflow":
+            logging.info("Unmask sflow service and start it")
+            duthost.shell("sudo systemctl unmask {}".format(container_name))
+            duthost.shell("sudo systemctl start {}".format(container_name))
+            # Sleep 10 seconds such that processes in sflow can come into live
+            time.sleep(10)
```

Comment:
This should not be specific to the sflow container. It should either be done for all containers which are disabled, or we should skip any disabled containers.